### PR TITLE
WIP - MySQL Database Schemas Feedback

### PR DESCRIPTION
**Pull request by awhigham9**
**This PR is a WORK IN PROGRESS (WIP) - I am opening it just to get feedback, not to complete it!**

# Description
This is a WIP PR to receive feedback regarding the current database schemas implemented as the MySQL adaptation of the existing MongoDB schema. This includes feedback on the schema overall and answers to specific questions regarding the intended design of the application data.

The primary questions are regarding the table schemas in `src/migration/buildDatabase.ts`. These questions are detailed below.
However, please feel free to provide any other feedback as well!

Some notes:
+ Disregard `src/migration/problem.ts`, it was just an example file being shared with others on the branch and will be removed.
+ The variables in `src/migration/env.example` will eventually have to make their way into the main env file(s), but are being kept separate now for simplicity.

### `Module` Table

- Should Number be the primary key, or is `id` sufficient as in Mongo? This actually applies to all of the tables, is there an application specific primary key or should a hidden `id` still be used?
- Should (Name, Number) form a composite key, and thus be constrained `UNIQUE`?

### `Problem` Table

- Should the `code_id` attribute be mandatory (i.e. `NOT NULL`)?
    - This would introduce a troublesome cyclic foreign key dependency, but accurately captures the intended 1:1 relationship between `Problem` and `Code`.
- Should any fields have defaults, such as time or memory limit?
- Should problem titles be unique?

### `Code` Table

- Should the attributes of the `Code` table be merged into `Problem` because the relationship
was intended to be 1:1?
- Alternatively, should we allow `Code(problem_id)` to be non-unique while not merging the tables? This would make it possible for there to be multiple code "objects" for a single Problem. This could be desirable in the future if Edugator will eventually support multiple languages for one problem. For example, you could solve a problem in C++ or Java.

# Related Issue
N/A for now (will open issues before real PRs later)

# Motivation and Context
These changes will eventually evolve to be the database tables storing Edugator's data in MySQL

# How Has This Been Tested?
A local MySQL installation has been used to validate the build creation's script correctness (Ubuntu 20.04.3 LTS within the WSL on Win10). CRUD on the data model has not yet been performed.

# Screenshots (if appropriate):
N/A

### CODE REVIEW

**Comment by awhigham9** [comment]
Also, the User table is a bit of a WIP. I added a `salt` attribute to store an user-unique integer salt to be used to hash passwords for storage. I am not sure if this is the best solution though, and MySQL might have some built in functionality of which I am not aware.

File: src/migration/buildDatabase.ts [hunk 1]
```diff
@@ -0,0 +1,147 @@
+import { createConnection, Connection } from 'mysql2';
```
Comment (marcgabe15) [inline src/migration/buildDatabase.ts:1]: Is there a way we can modularize this for unit testing? That way we can setup/teardown the database during unit tests?
Refere to this for mongo instance https://github.com/edugator-cise/edugator-staff-backend/blob/main/src/config/database.ts

File: src/migration/problem.orm.ts [hunk 2]
```diff
@@ -0,0 +1,66 @@
+import { ProblemInterface } from '../api/models/problem.model';
+import { Pool } from 'mysql2';
+import pool from './pool';
+
+// Is this possible without code duplication from ProblemInterface?
+interface ProblemQueryFilter {
```
Comment (marcgabe15) [inline src/migration/problem.orm.ts:6]: You can take out the two properties from ProblemInterface and then extend ProblemInterface to have those two values

File: src/migration/buildDatabase.ts (hunk 1 above)
Comment (awhigham9) [reply src/migration/buildDatabase.ts:1]: Yes, definitely. We can do something precisely like that file, it may just be a bit shorter.

**Comment by marcgabe15** [comment]
For the module quesitons:
- we only use number for ordering so id should be fine
- For the composite key, we could do that that just makes it so that (name,number) together are unique right?

**Comment by marcgabe15** [comment]
If there is a cyclic foreign key dependency then we can probably just have mySQLCode reference a problem id and omit code_id from the problem?

**Comment by marcgabe15** [comment]
For the code Table questions:
I like the idea of having non uniqueness that way we can have multipoe languages for each problem and add that extensibility

**Pull request closed without merging**